dev/build-tool-support/src/main/scala/com/lightbend/lagom/dev/PortAssigner.scala

```diff
@@ -34,59 +34,57 @@ object PortAssigner {
-            val port = findFirstAvailablePort(projectedPort, assignedPort)
-            loop(projectNames, assignedPort + port, unassigned.tail, result + (proj -> port))
-          case Nil => result
-          case proj +: rest =>
-            val projectedPort = projectedPortFor(proj)
-            if (assignedPort(projectedPort)) loop(rest, assignedPort, unassigned :+ proj, result)
-            else loop(rest, assignedPort + projectedPort, unassigned, result + (proj -> projectedPort))
-        }
-
-      loop(projectNamesWithTls.toSeq, Set.empty[Port], Vector.empty[ProjectName], Map.empty[ProjectName, Port])
+      if (enableSsl) lagomProjects.flatMap { plainName => Seq(plainName, plainName.withTls) }
+      else lagomProjects
+
+    val doubleMessage =
+      if (enableSsl) "The number of ports available must be at least twice the number of projects."
+      else ""
+
+    require(
+      projects.size <= range.delta,
+      s"""A larger port range is needed, as you have ${lagomProjects.size} Lagom projects and only ${range.delta}
+         |ports available. $doubleMessage
+         |You should increase the range passed for the lagomPortRange build setting.
+       """.stripMargin
+    )
+
+    @annotation.tailrec
+    def findFirstAvailablePort(port: Port, unavailable: Set[Port]): Port = {
+      // wrap around if the port's number equal the portRange max limit
+      if (!range.includes(port.value)) findFirstAvailablePort(Port(range.min), unavailable)
+      else if (unavailable(port)) findFirstAvailablePort(port.next, unavailable)
+      else port
     }
 
-    private def projectedPortFor(name: ProjectName): Port = {
+    @annotation.tailrec
+    def loop(projectNames: Seq[ProjectName], assignedPort: Set[Port], unassigned: Vector[ProjectName], result: Map[ProjectName, Port]): Map[ProjectName, Port] =
+      projectNames match {
+        case Nil if unassigned.nonEmpty =>
+          // if we are here there are projects with colliding hash that still need to get their port assigned. As expected, this step is carried out after assigning
+          // a port to all non-colliding projects.
+          val proj = unassigned.head
+          val projectedPort = projectedPortFor(proj)
+          val port = findFirstAvailablePort(projectedPort, assignedPort)
+          loop(projectNames, assignedPort + port, unassigned.tail, result + (proj -> port))
+        case Nil => result
+        case proj +: rest =>
+          val projectedPort = projectedPortFor(proj)
+          if (assignedPort(projectedPort)) loop(rest, assignedPort, unassigned :+ proj, result)
+          else loop(rest, assignedPort + projectedPort, unassigned, result + (proj -> projectedPort))
+      }
+
+    def projectedPortFor(name: ProjectName): Port = {
       val hash = Math.abs(name.hashCode())
       val portDelta = hash % range.delta
       Port(range.min + portDelta)
     }
+
+    loop(projects.toSeq, Set.empty[Port], Vector.empty[ProjectName], Map.empty[ProjectName, Port])
```

Comment:
Rename `loop` to `assignProjectPort` or something with a clearer name.

Reply:
Sure. (This name isn't new, I just un-nested it from `Project2PortMapBuilder`.)

Reply:
Yeah, notice that. Suggested because is not part of the API so we can change it to a better name.